### PR TITLE
Image layers

### PR DESCRIPTION
**Pull request by akpsgit**

### CODE REVIEW

File: clair/api.go [hunk 1]
```diff
@@ -107,6 +107,10 @@ func (a *apiV1) Analyze(image *docker.Image) ([]*Vulnerability, error) {
 			//since we are appending a pointer to the variable to the slice, we need to create a copy of the struct
 			//otherwise the slice winds up with multiple pointers to the same struct
 			vulnerability := v
+			if vulnerability.Metadata == nil {
+				vulnerability.Metadata = make(map[string]interface{})
+			}
```
Comment (FrimIdan) [inline clair/api.go]: why is it needed?

File: clair/api.go [hunk 2]
```diff
@@ -107,6 +107,10 @@ func (a *apiV1) Analyze(image *docker.Image) ([]*Vulnerability, error) {
...
+			vulnerability.Layer = strings.TrimPrefix(f.AddedBy, docker.TrimDigest(image.Digest))
```
Comment (FrimIdan) [inline clair/api.go]: can you please add comment/example what is being done here?

File: clair/clair.go [hunk 3]
```diff
@@ -55,6 +55,7 @@ type Vulnerability struct {
 	FixedIn        []feature              `json:"FixedIn,omitempty"`
 	FeatureName    string                 `json:"FeatureName",omitempty`
 	FeatureVersion string                 `json:"FeatureVersion",omitempty`
+	Layer          string                 `json:"Layer",omitempty`
```
Comment (FrimIdan) [inline clair/clair.go]: LayerHash?

File: docker/docker.go [hunk 4]
```diff
@@ -267,6 +289,131 @@ func (i *Image) Pull() error {
...
+// FetchFsCommands retrieves information about image layers commands from docker registry.
+func (i *Image) FetchFsCommands() error {
+	if i.FsCommands != nil {
+		log.Infof("Layer commands are already present")
+		return nil
+	}
+	ctx := context.Background()
+
+	opts := fanal_types.DockerOption{
+		Timeout:  90*time.Second,
```
Comment (FrimIdan) [inline docker/docker.go]: `NonSSL` should be added to support `REGISTRY_INSECURE` env var.
Also `InsecureSkipTLSVerify` to support `DOCKER_INSECURE`

File: docker/docker.go [hunk 5]
```diff
@@ -267,6 +289,131 @@ func (i *Image) Pull() error {
...
+	confB, err := json.Marshal(conf)
+	if err != nil {
+		return fmt.Errorf("failed to marshal config: %v", err)
+	}
+	log.Debugf("Image config: %s", confB)
+
+	var commands []string
+	for i, layerHistory := range conf.History {
+		if layerHistory.EmptyLayer {
+			log.Infof("Skipping empty layer (%v): %+v", i, layerHistory)
```
Comment (FrimIdan) [inline docker/docker.go:330]: Didn't we wanted to also show empty layers?

File: docker/docker.go [hunk 6]
```diff
@@ -267,6 +289,131 @@ func (i *Image) Pull() error {
...
+		return nil, func() {}, xerrors.Errorf("failed to parse the image name: %w", err)
+	}
+
+	//// Try accessing Docker Daemon
+	//img, cleanup, err := daemon.Image(ref)
+	//if err == nil {
+	//	// Return v1.Image if the image is found in Docker Engine
+	//	return img, cleanup, nil
+	//}
+	//result = multierror.Append(result, err)
```
Comment (FrimIdan) [inline docker/docker.go]: Do we need this?

File: docker/docker.go [hunk 7]
```diff
@@ -279,21 +426,32 @@ func parseImageResponse(resp *http.Response, image *Image) error {
...
 		}
-		image.digest = imageV2.Config.Digest
+		image.Digest = imageV2.Config.Digest
 		image.schemaVersion = imageV2.SchemaVersion
 	case "application/vnd.docker.distribution.manifest.v1+prettyjws":
-		var imageV1 imageV1
-		if err := json.NewDecoder(resp.Body).Decode(&imageV1); err != nil {
-			fmt.Fprintln(os.Stderr, "ImageV1 decode error")
-			return err
+		body, err := ioutil.ReadAll(resp.Body)
```
Comment (FrimIdan) [inline docker/docker.go:425]: check err?

File: docker/docker.go [hunk 8]
```diff
@@ -279,21 +426,32 @@ func parseImageResponse(resp *http.Response, image *Image) error {
...
+		// in schemaVersion 1 layers and commands are in reverse order, so we save them in the same order as v2
 		// base layer is the first
-		for i := range imageV1.FsLayers {
-			image.FsLayers[len(imageV1.FsLayers)-1-i].BlobSum = imageV1.FsLayers[i].BlobSum
+		for i := range schema1.FSLayers {
+			image.FsLayers[len(schema1.FSLayers)-1-i].BlobSum = schema1.FSLayers[i].BlobSum.String()
+			image.FsCommands[len(schema1.FSLayers)-1-i] = &FsLayerCommand{
+				Command: strings.Join(schema1.ExtractedV1Compatibility[len(schema1.FSLayers)-1-i].ContainerConfig.Cmd,","),
+				Layer:   schema1.FSLayers[i].BlobSum.Hex(),
+			}
```
Comment (FrimIdan) [inline docker/docker.go]: extract to func and unit test, it's not so trivial I think.

File: main.go [hunk 9]
```diff
@@ -1,16 +1,22 @@
 package main
 
 import (
+	//	"bytes"
+	"encoding/json"
+	"github.com/Portshift/klar/config"
+	"github.com/Portshift/klar/format"
+	"github.com/Portshift/klar/run"
+
+	//"encoding/json"
```
Comment (FrimIdan) [inline main.go]: please remove comments and sort imports

File: main.go [hunk 10]
```diff
@@ -88,10 +62,18 @@ func main() {
...
+	result.LayerCommands = commands
 	result.Success = true
 
+	resultB, err := json.Marshal(result)
+	if err != nil {
+		log.Errorf("Failed to marshal result: %v", err)
+		os.Exit(2)
+	}
+	log.Infof("resultB: %s", resultB)
+
```
Comment (FrimIdan) [inline main.go]: Do we need this?

File: docker/docker.go [hunk 11]
```diff
@@ -6,8 +6,14 @@ import (
 	"encoding/json"
 	"fmt"
 	"github.com/Portshift/klar/docker/token"
+	fanal_token "github.com/aquasecurity/fanal/image/token"
```
Comment (FrimIdan) [inline docker/docker.go]: sort imports

File: clair/api.go (hunk 1 above)
Comment (akpsgit) [reply clair/api.go]: not needed

File: docker/docker.go (hunk 6 above)
Comment (akpsgit) [reply docker/docker.go]: no :)

File: docker/docker.go (hunk 5 above)
Comment (akpsgit) [reply docker/docker.go:330]: I think that not currently from talking with Rafi. We can add this as part of layer support in dockle

File: main.go (hunk 10 above)
Comment (akpsgit) [reply main.go]: deleted

File: docker/docker.go (hunk 11 above)
Comment (akpsgit) [reply docker/docker.go]: same after gofmt

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+const (
+	OptionClairOutput        = "CLAIR_OUTPUT"
+	OptionClairAddress       = "CLAIR_ADDR"
+	OptionKlarTrace          = "KLAR_TRACE"
+	OptionClairThreshold     = "CLAIR_THRESHOLD"
+	OptionClairTimeout       = "CLAIR_TIMEOUT"
+	OptionDockerTimeout      = "DOCKER_TIMEOUT"
+	OptionJSONOutput         = "JSON_OUTPUT" // deprecate?
+	OptionFormatOutput       = "FORMAT_OUTPUT"
```
Comment (Erezf-p) [inline config/config.go:27]: not used

File: config/config.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/Portshift/klar/docker"
+	"github.com/Portshift/klar/utils"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+//Used to represent the structure of the whitelist YAML file
+type vulnerabilitiesWhitelistYAML struct {
```
Comment (Erezf-p) [inline config/config.go:14]: not used

File: docker/docker.go (hunk 8 above)
Comment (akpsgit) [reply docker/docker.go]: Thanks, this caught a bug:
Should be: strings.Join(schema1.ExtractedV1Compatibility[i]

File: go.mod [hunk 14]
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/opencontainers/go-digest v1.0.0-rc1
 	github.com/sirupsen/logrus v1.6.0
 	golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543
+	gotest.tools v2.2.0+incompatible
```
Comment (FrimIdan) [inline go.mod]: please remove - not used

**Review by FrimIdan — APPROVED** [review]

File: config/config.go (hunk 12 above)
Comment (akpsgit) [reply config/config.go:27]: This is part of the current klar (to be used from outside to print). I just moved the file

File: config/config.go (hunk 13 above)
Comment (akpsgit) [reply config/config.go:14]: This is part of the current klar (to be used from outside to print). I just moved the file

**Pull request merged**